src/witness/yamlWitness.ml

```diff
@@ -0,0 +1,114 @@
+open Analyses
+
+let uuid_random_state = Random.State.make_self_init ()
+
+module Make
+    (Spec : Spec)
+    (EQSys : GlobConstrSys with module LVar = VarF (Spec.C)
+                        and module GVar = GVarF (Spec.V)
+                        and module D = Spec.D
+                        and module G = Spec.G)
+    (LHT : BatHashtbl.S with type key = EQSys.LVar.t)
+    (GHT : BatHashtbl.S with type key = EQSys.GVar.t) =
+struct
+
+  module NH = BatHashtbl.Make (Node)
+
+  (* copied from Constraints.CompareNode *)
+  let join_contexts (lh: Spec.D.t LHT.t): Spec.D.t NH.t =
+    let nh = NH.create 113 in
+    LHT.iter (fun (n, _) d ->
+        let d' = try Spec.D.join (NH.find nh n) d with Not_found -> d in
+        NH.replace nh n d'
+      ) lh;
+    nh
+
+  let write lh gh =
+    (* yaml_conf is too verbose *)
+    (* let yaml_conf: Yaml.value = Json_repr.convert (module Json_repr.Yojson) (module Json_repr.Ezjsonm) (!GobConfig.json_conf) in *)
+    let yaml_creation_time = `String (TimeUtil.iso8601_now ()) in
+    let yaml_producer = `O [
+        ("name", `String "Goblint");
+        ("version", `String Version.goblint);
+        (* TODO: configuration *)
+        (* ("configuration", yaml_conf); *) (* yaml_conf is too verbose *)
+        ("command_line", `String Goblintutil.command_line);
+        (* TODO: description *)
+      ]
+    in
+    let files = GobConfig.get_string_list "files" in
+    let sha256_file f = Sha256.(to_hex (file f)) in
+    let sha256_file_cache = BatCache.make_ht ~gen:sha256_file ~init_size:5 in
+    let sha256_file = sha256_file_cache.get in
+    let yaml_task = `O ([
+        ("input_files", `A (List.map Yaml.Util.string files));
+        ("input_file_hashes", `O (List.map (fun file ->
+            (file, `String (sha256_file file))
+          ) files));
+        ("data_model", `String (match GobConfig.get_string "exp.architecture" with
+          | "64bit" -> "LP64"
+          | "32bit" -> "ILP32"
+          | _ -> failwith "invalid architecture"));
+        ("language", `String "C");
+      ] @ match !Svcomp.task with
+            | Some (module Task) -> [
+                ("specification", `String (Svcomp.Specification.to_string Task.specification))
+              ]
+            | None ->
+              []
+      )
+    in
+
+    let nh = join_contexts lh in
+
+    let yaml_entries = NH.fold (fun n local acc ->
+        match n with
+        | Statement _ ->
+          let context: Invariant.context = {
+              scope=Node.find_fundec n;
+              i = -1;
+              lval=None;
+              offset=Cil.NoOffset;
+              deref_invariant=(fun _ _ _ -> Invariant.none) (* TODO: should throw instead? *)
+            }
+          in
+          begin match Spec.D.invariant context local with
+            | Some inv ->
+              let inv = InvariantCil.exp_replace_original_name inv in
```

Comment:
You might want to do something like `pullOutCommonConjuncts` as I did in the assert transformation here?

Reply:
The issue is that the asserts become super-unreadable as soon as you have multiple paths, and pulling common conjuncts out makes it a lot nicer to inspect.

Reply:
Yup, I just realized I'll need to do it anyway to work around https://github.com/goblint/cil/issues/94.